### ISSUE TITLE
pykan 0.2.1 : Some errors

### ISSUE THREAD

**Issue by 911569318** [KindXiaoming/pykan#343]
## 1, Running hellokan produces the following error
<img width="990" alt="hellokan" src="https://github.com/user-attachments/assets/21f05139-434e-4809-a098-2ad9a6a59bf1">

## 2, Running helloMultKAN produces the following error
<img width="958" alt="MultKAN" src="https://github.com/user-attachments/assets/345c7e5e-caf2-4782-bd43-e07bae184c13">

## 3, Running Example 9: Singularity productes the following error
<img width="908" alt="Singularity" src="https://github.com/user-attachments/assets/2d6081c6-a5d7-41ce-9db3-d35b4cbd3076">

## 4, Attempt to create kan model with device=cuda:0
<img width="913" alt="cuda0" src="https://github.com/user-attachments/assets/16a0ffd8-e6df-4252-ad0f-08d601b306ec">

- It works as of version 0.0.5

## 5, Function with singularit
- By the way, Another question I would like to ask is why it is so difficult to deal with the regression problem of ln(x)?
<img width="899" alt="ln(x)" src="https://github.com/user-attachments/assets/cf992fc2-c451-4e5c-86fd-f428ade350a3">

## environments
- Here is the python environment I'm using, python version 3.9.7

 Name                    Version                   Build  Channel
anyio                     4.4.0                    pypi_0    pypi
argon2-cffi               23.1.0                   pypi_0    pypi
argon2-cffi-bindings      21.2.0                   pypi_0    pypi
arrow                     1.3.0                    pypi_0    pypi
asttokens                 2.4.1                    pypi_0    pypi
async-lru                 2.0.4                    pypi_0    pypi
attrs                     23.2.0                   pypi_0    pypi
babel                     2.15.0                   pypi_0    pypi
beautifulsoup4            4.12.3                   pypi_0    pypi
bleach                    6.1.0                    pypi_0    pypi
ca-certificates           2024.7.2             haa95532_0
certifi                   2024.7.4                 pypi_0    pypi
cffi                      1.16.0                   pypi_0    pypi
charset-normalizer        3.3.2                    pypi_0    pypi
colorama                  0.4.6                    pypi_0    pypi
comm                      0.2.2                    pypi_0    pypi
contourpy                 1.2.1                    pypi_0    pypi
cycler                    0.12.1                   pypi_0    pypi
debugpy                   1.8.2                    pypi_0    pypi
decorator                 5.1.1                    pypi_0    pypi
defusedxml                0.7.1                    pypi_0    pypi
exceptiongroup            1.2.2                    pypi_0    pypi
executing                 2.0.1                    pypi_0    pypi
fastjsonschema            2.20.0                   pypi_0    pypi
filelock                  3.13.1                   pypi_0    pypi
fonttools                 4.53.1                   pypi_0    pypi
fqdn                      1.5.1                    pypi_0    pypi
fsspec                    2024.2.0                 pypi_0    pypi
h11                       0.14.0                   pypi_0    pypi
httpcore                  1.0.5                    pypi_0    pypi
httpx                     0.27.0                   pypi_0    pypi
idna                      3.7                      pypi_0    pypi
importlib-metadata        8.0.0                    pypi_0    pypi
ipykernel                 6.29.5                   pypi_0    pypi
ipython                   8.18.1                   pypi_0    pypi
ipywidgets                8.1.3                    pypi_0    pypi
isoduration               20.11.0                  pypi_0    pypi
jedi                      0.19.1                   pypi_0    pypi
jinja2                    3.1.3                    pypi_0    pypi
joblib                    1.4.2                    pypi_0    pypi
json5                     0.9.25                   pypi_0    pypi
jsonpointer               3.0.0                    pypi_0    pypi
jsonschema                4.23.0                   pypi_0    pypi
jsonschema-specifications 2023.12.1                pypi_0    pypi
jupyter                   1.0.0                    pypi_0    pypi
jupyter-client            8.6.2                    pypi_0    pypi
jupyter-console           6.6.3                    pypi_0    pypi
jupyter-core              5.7.2                    pypi_0    pypi
jupyter-events            0.10.0                   pypi_0    pypi
jupyter-lsp               2.2.5                    pypi_0    pypi
jupyter-server            2.14.2                   pypi_0    pypi
jupyter-server-terminals  0.5.3                    pypi_0    pypi
jupyterlab                4.2.3                    pypi_0    pypi
jupyterlab-pygments       0.3.0                    pypi_0    pypi
jupyterlab-server         2.27.2                   pypi_0    pypi
jupyterlab-widgets        3.0.11                   pypi_0    pypi
kiwisolver                1.4.5                    pypi_0    pypi
markupsafe                2.1.5                    pypi_0    pypi
matplotlib                3.6.2                    pypi_0    pypi
matplotlib-inline         0.1.7                    pypi_0    pypi
mistune                   3.0.2                    pypi_0    pypi
mpmath                    1.3.0                    pypi_0    pypi
nbclient                  0.10.0                   pypi_0    pypi
nbconvert                 7.16.4                   pypi_0    pypi
nbformat                  5.10.4                   pypi_0    pypi
nest-asyncio              1.6.0                    pypi_0    pypi
networkx                  3.2.1                    pypi_0    pypi
notebook                  7.2.1                    pypi_0    pypi
notebook-shim             0.2.4                    pypi_0    pypi
numpy                     1.24.4                   pypi_0    pypi
openssl                   1.1.1w               h2bbff1b_0
overrides                 7.7.0                    pypi_0    pypi
packaging                 24.1                     pypi_0    pypi
pandas                    2.2.2                    pypi_0    pypi
pandocfilters             1.5.1                    pypi_0    pypi
parso                     0.8.4                    pypi_0    pypi
pillow                    10.4.0                   pypi_0    pypi
pip                       24.0             py39haa95532_0
platformdirs              4.2.2                    pypi_0    pypi
prometheus-client         0.20.0                   pypi_0    pypi
prompt-toolkit            3.0.47                   pypi_0    pypi
psutil                    6.0.0                    pypi_0    pypi
pure-eval                 0.2.2                    pypi_0    pypi
pycparser                 2.22                     pypi_0    pypi
pygments                  2.18.0                   pypi_0    pypi
pykan                     0.2.1                    pypi_0    pypi
pyparsing                 3.1.2                    pypi_0    pypi
python                    3.9.7                h6244533_1
python-dateutil           2.9.0.post0              pypi_0    pypi
python-json-logger        2.0.7                    pypi_0    pypi
pytz                      2024.1                   pypi_0    pypi
pywin32                   306                      pypi_0    pypi
pywinpty                  2.0.13                   pypi_0    pypi
pyyaml                    6.0.1                    pypi_0    pypi
pyzmq                     26.0.3                   pypi_0    pypi
qtconsole                 5.5.2                    pypi_0    pypi
qtpy                      2.4.1                    pypi_0    pypi
referencing               0.35.1                   pypi_0    pypi
requests                  2.32.3                   pypi_0    pypi
rfc3339-validator         0.1.4                    pypi_0    pypi
rfc3986-validator         0.1.1                    pypi_0    pypi
rpds-py                   0.19.0                   pypi_0    pypi
scikit-learn              1.1.3                    pypi_0    pypi
scipy                     1.13.1                   pypi_0    pypi
send2trash                1.8.3                    pypi_0    pypi
setuptools                65.5.0                   pypi_0    pypi
six                       1.16.0                   pypi_0    pypi
sniffio                   1.3.1                    pypi_0    pypi
soupsieve                 2.5                      pypi_0    pypi
sqlite                    3.45.3               h2bbff1b_0
stack-data                0.6.3                    pypi_0    pypi
sympy                     1.11.1                   pypi_0    pypi
terminado                 0.18.1                   pypi_0    pypi
threadpoolctl             3.5.0                    pypi_0    pypi
tinycss2                  1.3.0                    pypi_0    pypi
tomli                     2.0.1                    pypi_0    pypi
torch                     2.2.2+cu121              pypi_0    pypi
torchaudio                2.2.2+cu121              pypi_0    pypi
torchvision               0.17.2+cu121             pypi_0    pypi
tornado                   6.4.1                    pypi_0    pypi
tqdm                      4.66.2                   pypi_0    pypi
traitlets                 5.14.3                   pypi_0    pypi
types-python-dateutil     2.9.0.20240316           pypi_0    pypi
typing-extensions         4.9.0                    pypi_0    pypi
tzdata                    2024.1                   pypi_0    pypi
uri-template              1.3.0                    pypi_0    pypi
urllib3                   2.2.2                    pypi_0    pypi
vc                        14.2                 h2eaa2aa_4
vs2015_runtime            14.29.30133          h43f2093_4
wcwidth                   0.2.13                   pypi_0    pypi
webcolors                 24.6.0                   pypi_0    pypi
webencodings              0.5.1                    pypi_0    pypi
websocket-client          1.8.0                    pypi_0    pypi
wheel                     0.43.0           py39haa95532_0
widgetsnbextension        4.0.11                   pypi_0    pypi
zipp                      3.19.2                   pypi_0    pypi

**Comment by KindXiaoming** [comment]
1. model = KAN(..., auto_save=True)
2. the latest version should already solved this.
3. model.fit(..., singularity_avoiding=True)
4. yes this is a bug to be resolved
5. it's actually working well but not simplified.

**Comment by 911569318** [comment]
> 1. model = KAN(..., auto_save=True)
> 2. the latest version should already solved this.
> 3. model.fit(..., singularity_avoiding=True)
> 4. yes this is a bug to be resolved
> 5. it's actually working well but not simplified.
Thank you for your quick reply. Most of the problems have been solved except for error 2. I noticed some new features (such as singularity_avoiding=True) that were very useful for my work. Thanks again.

**Comment by 911569318** [comment]
> 5. it's actually working well but not simplified.
It was a good attempt. Ln(x) is very commonly used in statistical physics. I am looking forward to KANs's improvement in this area.
![image](https://github.com/user-attachments/assets/a62ec127-c44e-4993-9922-461ffb6f59c5)

**Comment by KindXiaoming** [comment]
could you please say more what is Ln(x)? :)

**Comment by 911569318** [comment]
> could you please say more what is Ln(x)? :)

Sorry, maybe I'm not being precise. I mean log of x, base e of log x.

**Comment by 911569318** [comment]
> 2. the latest version should already solved this.

Error 2 has been resolved. I downloaded the kan folder from github and used it to replace the kan folder that pypi automatically installed, and it worked.Thanks again for your reply
